src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+                   push_code_page(mapping_aliases, grapheme) ||
+                   push_fallback(grapheme);
+        };
+
+        if (push(grapheme))
+            continue;
+
+        if (grapheme.HasMark()) {
+            grapheme.StripMarks();
+            if (push(grapheme))
+                continue;
+        }
+
+        push_unknown(grapheme.GetCodePoint());
+    }
+}
+
+// ***************************************************************************
+// Read resources from files
+// ***************************************************************************
+
+static bool prepare_code_page(const uint16_t code_page);
+
+template <typename T1, typename T2>
+bool add_if_not_mapped(std::map<T1, T2> &mapping, T1 first, T2 second)
+{
+    if (mapping.count(first))
+        return false;
+
+    mapping[first] = second;
+    return true;
+}
+
+static std::ifstream open_mapping_file(const std_fs::path &path_root,
+                                       const std::string &file_name)
+{
+    const std_fs::path file_path = path_root / file_name;
+    std::ifstream in_file(file_path.string());
+    if (!in_file)
+        LOG_ERR("UTF8: Could not open mapping file %s", file_name.c_str());
+
+    return in_file;
+}
+
+static bool get_line(std::ifstream &in_file, std::string &line_str, size_t &line_num)
+{
+    line_str.clear();
+
+    while (line_str.empty()) {
+        if (!std::getline(in_file, line_str))
+            return false;
+        if (line_str.length() >= 1 && line_str[0] == end_of_file_marking)
+            return false; // end of definitions
+        ++line_num;
+    }
+
+    return true;
+}
+
+static bool get_tokens(const std::string &line, std::vector<std::string> &tokens)
+{
+    // Split the line into tokens, strip away comments
+
+    bool token_started = false;
+    size_t start_pos   = 0;
+
+    tokens.clear();
+    for (size_t i = 0; i < line.size(); ++i) {
+        if (line[i] == '#')
+            break; // comment started
+
+        const bool is_space = (line[i] == ' ')  || (line[i] == '\t') ||
+                              (line[i] == '\r') || (line[i] == '\n');
+
+        if (!token_started && !is_space) {
+            // We are at first character of the token
+            token_started = true;
+            start_pos     = i;
+            continue;
+        }
+
+        if (token_started && is_space) {
+               // We are at the first character after the token
+            token_started = false;
+            tokens.emplace_back(line.substr(start_pos, i - start_pos));
+            continue;
+        }
+    }
+
+    if (token_started)
+        // We have a token which ends with the end of the line
+        tokens.emplace_back(line.substr(start_pos));
+
+   return !tokens.empty();
+}
+
+static bool is_hex_digit(const char digit)
+{
+    return (digit >= '0' && digit <= '9') ||
+           (digit >= 'a' && digit <= 'f') ||
+           (digit >= 'A' && digit <= 'F');
+}
+
```

Comment:
Not sure if `isxdigit` can fill this need? <https://cplusplus.com/reference/cctype/isxdigit/>

Reply:
Yes, this one will do. Changed.